clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -576,12 +576,8 @@ public void awaitFlushCompletion() throws InterruptedException {
         }
     }
 
-    public boolean hasUnflushedBatches() {
-        for (Map.Entry<TopicPartition, Deque<ProducerBatch>> entry : this.batches().entrySet()) {
-            if (!entry.getValue().isEmpty())
-                return true;
-        }
-        return !this.incomplete.incomplete.isEmpty();
+    public boolean hasIncompleteBatches() {
+        return !this.incomplete.isEmpty();
```

Comment:
As far as I can tell, checking `incomplete` is sufficient to handle both sent and unsent batches.